Xamarin.Forms.Core/Markup/GridRowColEnums.cs

```diff
@@ -0,0 +1,64 @@
+﻿using System;
+using System.Globalization;
+
+namespace Xamarin.Forms.Markup
+{
+	public static class GridRowColEnums
+	{
+		public static GridLength Auto => GridLength.Auto;
+
+		public static GridLength Star => GridLength.Star;
+
+		public static class Columns
+		{
+			public static ColumnDefinitionCollection Define<TEnum>(params (TEnum name, GridLength width)[] cols) where TEnum : Enum
+			{
+				var columnDefinitions = new ColumnDefinitionCollection();
+				for (int i = 0; i < cols.Length; i++)
+				{
+					if (i != cols[i].name.ToInt())
+						throw new ArgumentException(
```

Comment:
This exception serves to ensure that the _visual order_ of the rows in Rows.Define remains identical to the _runtime order_, especially when editing rows in complex layouts. Sorting would promote inconsistency between source and runtime layout.

Even when not using enums, devs tend to order the row definitions (and child controls) in their grids on ascending row number, so the markup source gives them a mental image consistent with runtime. It is good for maintainability. Since enums can easily change value and the value is not visible in the markup as with numbers, devs could more easily overlook a mismatch; this exception helps to signal mismatches in a hard-to-miss manner at the earliest possible moment.

It would be more friendly to signal this inconsistency at build time, but that would require VS build / analyze tooling (maybe something for the roadmap?). Since this exception will always be thrown when the Grid is built, there is no risk of the exception going unfixed to production.

I considered making this a trace error, but in my experience that is very easy to miss in an early stage and therefore would mean more expensive maintenance later in the DevOps process (or no fix at all).

Make sense?

Reply:
I'd vote for sorting, XAML doesn't make you define them in order so why should the fluent API.  I agree it's better to keep them ordered in code or XAML the same as display, but I sometimes re-order rows, etc while working with the layout and then clean it up when I'm done.  Quicker to just change a row or column when iterating.  Plus it would be more consistent with XAML and plain C# to allow any ordering.

Reply:
@bmacombe 
> XAML doesn't make you define them in order

Afaik the _only_ way to define rows in XAML is in order?

```XAML
  <Grid.RowDefinitions>
    <RowDefinition Height="2*" />
    <RowDefinition Height="*" />
    <RowDefinition Height="200" />
  </Grid.RowDefinitions>
```
The check on order only becomes necessary because we use enums as indexes.
So the exception ensures that fluent is consistent with XAML, while a sort would make it inconsistent.

Reply:
@VincentH-Net My mistake then, I didn't look at the code close enough...I was thinking assigning rows and columns to child controls of the grid.

Reply:
@bmacombe NP, I suspected as much. Thx for giving feedback!

Reply:
@hartez I think we should leave this exception in, to prevent introducing markup errors that only have become possible by using enums (at least until we have edit / build time tooling that can do this).

Also see above:

> the exception ensures that fluent is consistent with XAML, while a sort would make it inconsistent